File: Scheduling Algos/fcfs/python/fcfs-no-sort.py
def findCompletionTime():
    fin = [False] * len(proc)
    time, idx = 0, 0
    end, lastArrival = 0, max(proc, key = lambda x: x['at'])['at'] + 1
    while end < len(proc):
        curr = lastArrival
        for elem in zip(fin, proc):
            if not elem[0] and curr > elem[1]['at']:
                idx = elem[1]['pid']
                curr = elem[1]['at']
        if time < curr:
            time = curr
        fin[idx] = True
        time += proc[idx]['bt']
        proc[idx]['ct'] = time
        end += 1

def findAvgTime():
    print('Avg TAT: ', end = '')
    print('%.3f'%(sum((p['ct'] - p['at']) for p in proc) / len(proc)))
    print('Avg WT: ', end = '')
    print('%.3f'%(sum((p['ct'] - (p['at'] + p['bt'])) for p in proc) / len(proc)))

if __name__ == '__main__':
    proc = [
            {'pid': 0, 'at': 2, 'bt': 5, 'ct': 0},
            {'pid': 1, 'at': 0, 'bt': 3, 'ct': 0},
            {'pid': 2, 'at': 1, 'bt': 6, 'ct': 0},
            {'pid': 3, 'at': 2, 'bt': 2, 'ct': 0},
            {'pid': 4, 'at': 7, 'bt': 1, 'ct': 0},
            {'pid': 5, 'at': 5, 'bt': 2, 'ct': 0}
            ]
    findCompletionTime()

    print('PID\tAT\tBT\tCT\tTAT\tWT')
    [print(f"{p['pid']}",
        f"{p['at']}",
        f"{p['bt']}",
        f"{p['ct']}",
        f"{p['ct'] - p['at']}",
        f"{p['ct'] - (p['at'] + p['bt'])}",
        sep = '\t') for p in proc]
    findAvgTime()

File: algorithms/fcfs/cpp/stl/fcfs-no-sort.cpp
#include <iostream>
#include <vector>
#include <algorithm>

struct Process {
	unsigned int pid;
	unsigned int at;
	unsigned int bt;
	unsigned int ct;
};

// ------------------------------------------------------
// Finding the waiting and turn around time, after computing the
// completion time.

void findCompletionTime(std::vector<Process> &proc) {
	unsigned count = proc.size();
	std::vector<bool> fin(count, true);
	unsigned end = 0, time = 0;
	unsigned idx = 0;
	unsigned lastArrival = std::max_element(
			proc.begin(),
			proc.end(),
			[] (const Process &a, const Process &b) -> bool {
			return a.at < b.at;
			})->at + 1;
	while(end < count) {
		unsigned min = lastArrival;
		for(auto &p: proc)
			if (fin[p.pid] && min > p.at) {
				min = p.at;
				idx = p.pid;
			}
		if (time < min)
			time = min;
		fin[idx] = false;
		time += proc[idx].bt;
		proc[idx].ct = time;
		end++;
	}
}

void findAvgTime(std::vector<Process> &proc) {
	findCompletionTime(proc);

	float ta{0}, w{0};

	std::cout << "PID\tAT\tBT\tCT\tTAT\tWT\n";
	for (auto &p: proc) {
		unsigned pTA = p.ct - p.at;
		unsigned pW = pTA - p.bt;
		std::cout << p.pid << '\t' << p.at << '\t'
			<< p.bt << '\t' << p.ct << '\t'
			<< pTA << '\t' << pW << '\n';
		ta += pTA;
		w += pW;
	}

	ta /= proc.size();
	w /= proc.size();

	std::cout << std::endl;
	std::cout << "Average Waiting Time: " << w << '\n' 
		<< "Average Turn-Around Time: " << ta << '\n';
}

// ------------------------------------------------------
// Driver for the program

int main() {
	std::vector<Process> proc = {
		{0, 5, 2, 0},
		{1, 5, 3, 0},
		{2, 1, 4, 0},
		{3, 2, 2, 0},
		{4, 1, 2, 0},
		{5, 1, 1, 0}
	};

	findAvgTime(proc);

	std::cout << '\n';
	return 0;
}
